### PR TITLE
fix(hybridcloud) Refactor and start cleanup on webhook bucketing

### PR DESCRIPTION
**Pull request by markstory**
Webhook bucketing has been working pretty well in production for the past few days. These changes move the common code up and add a new template method that parsers can implement to use bucketing.

I've added an additional cache key that is used to track bucketing mode being active over a longer period of time. With only the rate limit, we get a bundle of payloads in the main mailbox at the top of each hour. I'd prefer that we do that less often as it can cause out of order delivery.

### CODE REVIEW

File: src/sentry/middleware/integrations/parsers/base.py [hunk 1]
```diff
@@ -182,6 +185,42 @@ def get_response_from_webhookpayload_for_integration(
...
+        """
+        Used by integrations with higher hook volumes to create smaller mailboxes
+        that can be delivered in parallel. Requires the integration to implement
+        `mailbox_bucket_id`
+        """
+        # If we get fewer than 3000 in 1 hour we don't need to split into buckets
+        ratelimit_key = f"webhookpayload:{self.provider}:{integration.id}"
+        use_buckets_key = f"{ratelimit_key}:use_buckets"
+
+        use_buckets = cache.get(use_buckets_key)
```
Comment (GabeVillalobos) [inline src/sentry/middleware/integrations/parsers/base.py:198]: What's backing this cache? Is this redis or memcache by chance?

File: src/sentry/middleware/integrations/parsers/base.py [hunk 2]
```diff
@@ -182,6 +185,42 @@ def get_response_from_webhookpayload_for_integration(
...
         )
 
+    def get_mailbox_identifier(self, integration: RpcIntegration, data: Mapping[str, Any]) -> str:
+        """
+        Used by integrations with higher hook volumes to create smaller mailboxes
+        that can be delivered in parallel. Requires the integration to implement
+        `mailbox_bucket_id`
+        """
+        # If we get fewer than 3000 in 1 hour we don't need to split into buckets
+        ratelimit_key = f"webhookpayload:{self.provider}:{integration.id}"
```
Comment (GabeVillalobos) [inline src/sentry/middleware/integrations/parsers/base.py:195]: Is this used by our stripe webhook handling in any way? Curious if we'll encounter an issue with that at all.

**Review by GabeVillalobos — APPROVED** [review]
Looks good! Just a couple clarifying questions.

File: src/sentry/middleware/integrations/parsers/base.py (hunk 1 above)
Comment (markstory) [reply src/sentry/middleware/integrations/parsers/base.py:198]: This is memcache.

File: src/sentry/middleware/integrations/parsers/base.py (hunk 2 above)
Comment (markstory) [reply src/sentry/middleware/integrations/parsers/base.py:195]: It isn't. Parsers have to opt in to using `get_webhook_identifier` when creating outbox messages. So far stripe is always low volume and doesn't need additional partitioning.

**Pull request merged**